### PR TITLE
[SPARK-28601][CORE][SQL] Use StandardCharsets.UTF_8 instead of "UTF-8" string representation, and get rid of UnsupportedEncodingException

### PR DESCRIPTION
**Pull request by HeartSaVioR**
## What changes were proposed in this pull request?

This patch tries to keep consistency whenever UTF-8 charset is needed, as using `StandardCharsets.UTF_8` instead of using "UTF-8". If the String type is needed, `StandardCharsets.UTF_8.name()` is used.

This change also brings the benefit of getting rid of `UnsupportedEncodingException`, as we're providing `Charset` instead of `String` whenever possible.

This also changes some private Catalyst helper methods to operate on encodings as `Charset` objects rather than strings.

## How was this patch tested?

Existing unit tests.

### CODE REVIEW

**Comment by dongjinleekr** [comment]
+1 for this PR.

**Comment by gaborgsomogyi** [comment]
Even if I'm on vacation this looks simple enough to give a helping hand 🙂

Looks good but maybe 'import java.nio.charset.StandardCharsets.UTF_8' could be used all the places to save some characters.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala]: Shall we import `java.nio.charset.StandardCharsets.UTF_8` instead of splitting this into two lines?

File: sql/hive-thriftserver/v1.2.1/src/main/java/org/apache/hive/service/cli/operation/HiveCommandOperation.java
Comment (dongjoon-hyun) [inline sql/hive-thriftserver/v1.2.1/src/main/java/org/apache/hive/service/cli/operation/HiveCommandOperation.java]: ditto. Shall we import `java.nio.charset.StandardCharsets.UTF_8` instead?

File: sql/hive-thriftserver/v1.2.1/src/main/java/org/apache/hive/service/cli/operation/SQLOperation.java
Comment (dongjoon-hyun) [inline sql/hive-thriftserver/v1.2.1/src/main/java/org/apache/hive/service/cli/operation/SQLOperation.java]: Nice change!

File: sql/hive-thriftserver/v2.3.5/src/main/java/org/apache/hive/service/cli/operation/SQLOperation.java
Comment (dongjoon-hyun) [inline sql/hive-thriftserver/v2.3.5/src/main/java/org/apache/hive/service/cli/operation/SQLOperation.java]: Do we still need this `UnsupportedEncodingException`?

File: core/src/main/scala/org/apache/spark/executor/ProcfsMetricsGetter.scala
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/executor/ProcfsMetricsGetter.scala]: Ditto. Let's import `java.nio.charset.StandardCharsets.UTF_8` instead of splitting this into two lines?

File: core/src/main/scala/org/apache/spark/ui/PagedTable.scala
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/ui/PagedTable.scala]: ditt.

File: sql/hive-thriftserver/v2.3.5/src/main/java/org/apache/hive/service/cli/operation/SQLOperation.java
Comment (dongjoon-hyun) [reply sql/hive-thriftserver/v2.3.5/src/main/java/org/apache/hive/service/cli/operation/SQLOperation.java]: Ur, @HeartSaVioR . Could you test this again, please?
For me, it's used without a proper `import`. This will fail at compilation.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Could you build and test again, @HeartSaVioR ?
- `-Phadoop-3.2` might help your compilation test.
- After you passed the compilation, please remove invalid `catch (UnsupportedEncodingException e)` together, too.

File: core/src/main/scala/org/apache/spark/executor/ProcfsMetricsGetter.scala
Comment (srowen) [reply core/src/main/scala/org/apache/spark/executor/ProcfsMetricsGetter.scala]: Actually you can pass the `Charset` directly here, and avoid `name()`. That would be a little better. Same for any `InputStreamReader`

File: repl/src/main/scala/org/apache/spark/repl/ExecutorClassLoader.scala
Comment (dongjoon-hyun) [inline repl/src/main/scala/org/apache/spark/repl/ExecutorClassLoader.scala]: Ditto. Change `import` instead of splitting lines.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala]: I wonder if we can change these private methods to accept a `Charset`? it'd be tidier.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/sources/TextSocketStreamSuite.scala
Comment (srowen) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/sources/TextSocketStreamSuite.scala]: No need for `.name()`?

File: sql/hive-thriftserver/v2.3.5/src/main/java/org/apache/hive/service/cli/operation/SQLOperation.java
Comment (srowen) [reply sql/hive-thriftserver/v2.3.5/src/main/java/org/apache/hive/service/cli/operation/SQLOperation.java]: Yes, here and above I think this can no longer throw UnsupportedEncodingException

**Review by srowen — COMMENTED** [review]
I don't know how valuable just replacing the string with constant is, but it's OK.
The changes that cause methods to use `Charset` directly are a small win. It avoids looking up the charset and error checking.

**Comment by HeartSaVioR** [comment]
Thanks for the quick and detailed reviewing! Btw, is there any approach to enable build for both hadoop 2.7 and hadoop 3.2? Looks like this PR needs both and only `hadoop-2.7` profile is enabled in build.

**Comment by srowen** [comment]
I don't think the PR builder can test vs 3.2, but we have a separate job for that. I don't see what might be different for 3.2 though?

File: sql/hive-thriftserver/v2.3.5/src/main/java/org/apache/hive/service/cli/operation/HiveCommandOperation.java
Comment (HyukjinKwon) [inline sql/hive-thriftserver/v2.3.5/src/main/java/org/apache/hive/service/cli/operation/HiveCommandOperation.java]: I think he meant hadoop 3.2 due to this code path. Actually lately Yuming added a way to use hadoop 3.2 profile in PR builder as of https://github.com/apache/spark/pull/24045
It doesn't look quite likely hadoop 3.2 build matters though.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HeartSaVioR** [comment]
This patch touches various versions of hive-thrift server code which seems to require builds from both profiles (hadoop-2.7, hadoop-3.1).

**Comment by srowen** [comment]
Oh I see your point. Can you try compiling with the 3.2 profile locally as a check? I think it will be fine in any event.

**Comment by HyukjinKwon** [comment]
FWIW, I triggered the build against hadoop3.2 profile in this PR builder by changing its PR title.

```
Building Spark
========================================================================
[info] Building Spark using SBT with these arguments:  -Phadoop-3.2 -Pkubernetes -Phive-thriftserver -Phadoop-cloud -Pkinesis-asl -Pyarn -Pspark-ganglia-lgpl -Phive -Pmesos test:package streaming-kinesis-asl-assembly/assembly
...
```

**Comment by HeartSaVioR** [comment]
I've run both builds locally (if I remember correctly) before pushing the last commit, and we've run both builds for the last commit, so unless we touch more, looks like it's good to go.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala]: Ur, this is an exceptional function signature change which is different from the rest of this PR.
Please mention this clearly in the PR description with the following code snippets.
```
- def fileToString(file: File, encoding: String = "UTF-8"): String = {
+ def fileToString(file: File, encoding: Charset = UTF_8): String = {
```

Comment (srowen) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala]: I think the change is fine to this private method in this PR. If it's just the description in question, I can edit it.

**Comment by dongjoon-hyun** [comment]
Thanks for the fix, @srowen . The change is fine. I didn't ask to put it out as you saw [here](https://github.com/apache/spark/pull/25335#discussion_r310395522).

However, the reviewers had better give a clear direction to the contributors and help them make better PRs next time. 

Previously, I sometimes revised some PR descriptions by myself and informed them. And, it turns out that the contributors do not pay attentions on their next PR titles and descriptions more and more. That was not helpful for the community growth.

The problem I pointed in this PR was that this unintentionally disguised as an innocent PR replacing a string literal with the predefined one. (`function signature change` was not mentioned at all). This is very wrong because it makes the review harder. After merging this PR, the commit log will mislead the readers definitly.

That's the reason why I do nit-picking on the misleading PRs. Especially, PR title/description/patch consistency.

**Comment by HeartSaVioR** [comment]
I understand the intention and appreciate the continuous effort on improving quality of commit message. Thanks!

There're some points I honestly have 2 cents: 

1) The definition of 'good' varies to everyone. One example: now the commit title is far from guide in `git commit`, 50 characters vs 150+ characters. Even without tagging it exceeds 100 characters. 
When contributors may not agree the guide of revising commit title or description (when they are having different perspective of 'good') but just follow the guide and change the title/description to let PR just merge in, nothing would be changed in further PR. We shouldn't expect it even we put effort to educate them. If we think something is mandatory to be included in the PR's title or description, we should discuss and document it explicitly. (Like #25310 - actually I'm not sure I could agree requiring contributors even describe "developer-facing" change though. Happy to see it excluded.)

2) The project seems to go on direction for asking more and more efforts on contributors, even on minors. Time is gold, not only for reviewers, but also for contributors. We may need to find some balance here.

3) We might want to always remind that we are not doing like reviewing minors with microscope while skimming the code change for 1000+ of lines. So many developers agreed on this old tweet: https://twitter.com/iamdevloper/status/397664295875805184 (It would be fun to go through the thread.) I'd rather say "importance" should be considered instead of "lines", but minors are minors, both code lines and importance.

**Comment by HyukjinKwon** [comment]
I think we should update the guidelines to force something. If something doesn't comply to the guide line, we should fix. Otherwise, we call committer's judgement.

I agree @dongjoon-hyun's in general because I fix PR descriptions and titles too when it looks not well-formed by myself. I do nitpick too time to time but maybe we should avoid to nitpick too much.

**Comment by dongjoon-hyun** [comment]
I'm very sad to hear your complaint like the following that because I'm the one who tried to help you in several your PRs as a committer.
> The project seems to go on direction for asking more and more efforts on contributors, even on minors. Time is gold, not only for reviewers, but also for contributors. We may need to find some balance here.

@HeartSaVioR . The only reason why the committers comments on your PRs is to **merge** your PR quickly. Otherwise, I'll will not be here. As you know as an Apache Storm PMC, if a committer feels a PR is mergeable, he become active on that PR to complete it quickly. It's not to sabotage it.

**Comment by HeartSaVioR** [comment]
First of all, whatever my intention was, apologize my bad, I was pretty bad at representing my voices in formal way if you feel this as complaint. My 2 cents was not only for this PR but across all the mood in the project.

To be honest, I'd rather just follow whatever reviewers (especially someone with commit privilege) ask to fix it and don't say anything, if I just wanted to let commit be merged. Easiest way and no noise. If that's how the project works I would respect and just follow it.

What I pointed out is regarding some kind of preference which sometimes make contributors back and forth, as reviewers are not sharing consistent view. If we can see the values among community, we could reach a consensus in official discussion (either small or require votes) and require it. I'm afraid it can be educated, and it should be educated once someone feels it as preference. Honestly, the only way to correct it would be correcting time to time. Contributors can't read reviewers' mind in prior.

**Comment by srowen** [comment]
@dongjoon-hyun @HeartSaVioR My $0.02:

I do think it's ideal to keep the PR description up to date as it becomes part of the commit log. I don't know if it was a big deal in this case as the additional change was a minor internal refactoring, and we wouldn't exhaustively list every single change in the description. 

I don't see harm in the reminder or implementing the one-line edit, so it was fine. It was minor but not nit-picky. This is a judgment call and this is part of a process of synchronizing that judgment.

There's a very good point here that minor changes inevitably get more scrutiny because there's less to look at. That's not great but it's nature. If you view the occasional microscopic review as more an exercise in communicating about standards, and something that happens when the substance of the change is straightforward, it makes more sense. 

I'd also say we might review in more detail for people who are making many good contributions, as that person is growing into a reviewer too and perhaps on the path to committer, whereas it's not worth communicating this for a one-off contributor, and we might just let it go or fix it ourselves. Not great rationalizations, but it's not so bad.

You are right that reviewers don't even always agree, which is also just life. It can be frustrating if it's happening on your PR but is part of the process of synchronization. Reviewers do really need to bias towards finding consensus quickly.

- We might be able to emphasize these details in the PR template
- Reviewers should probably keep in mind the contributor's time, sure, and judge how much another round of minor changes 'costs' vs value
- Contributors may just have to put up with a bit of this process of synchronization sometimes and, weirdly, I hope feel more included as a result

**Comment by HeartSaVioR** [comment]
Thanks for the detailed voice, @srowen. I totally agree on your points.

I have to say again, my bad I admit I was a bit rushed on this: not all of my 2 cents apply to this PR but I made confusion that I'm complaint on these things for this PR. I have been struggled with getting review on major PRs for fairly long time and it still doesn't work (that's the main concern, honestly) so as a workaround I have to pick up minors hoping these minors would pass review phase easily, but I failed to isolate the root cause of my concern, and expressed concerns on the wrong place. Apologize to @dongjoon-hyun on this again. Sorry about that.

Likewise "synchronization of judgement" I still think providing consistent view among reviewers would be ideal to contributors. It's natural for reviewers to have different perspectives (sure I'm not blaming this. it's natural.), but once it's happening, ideally discussion should happen among reviewers (maybe including author as well if necessary) and guide as one direction to let contributors follow up without confusion. 

It's pretty hard to get rid of individual's preference while reviewing, but it's ideal to try out - asking ourselves again what's the current practice of this project for this guide. Trying to educate without some documented guide might bring unintended reaction - feeling the guide as required to follow one's preference - and it wouldn't gonna work, really. That's the reason style guide exists, PR template exists, guide doc exists, etc.

Btw, maybe this would help reducing efforts if we do update PR's title/description for what we changed during review phase when we are just about to merge. Syncing-up PR's title/description with code changes would be non-trivial effort, especially code changes is still happening. If back-and-forth happens, experience of contributors would be going pretty bad, has to fix both code and PR. At least we could try to avoid this.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to master.
Thank you all for sharing your thoughts.

Always people over code!

**Comment by HyukjinKwon** [comment]
LGTM too. Just noticed you guys here are all pretty senior in OSS :).

**Comment by HeartSaVioR** [comment]
Sorry again for the noise here, and thanks all for the detailed reviews, valuable feedbacks, and merge!

**Comment by srowen** [comment]
You raise a good point about not being able to get review attention sometimes @HeartSaVioR . Sometimes something isn't reviewed or merged because it's not obviously wrong, just not obviously worth it or important. That's a soft 'no' and I'd prefer to be more explicit about it, but sometimes that isn't received well.

Sometimes it's just that nobody has picked up that particular change to look at in the long stream of new issues each day, or the right person hasn't seen it. For example I honestly ignore SQL changes and focus on ML, core, docs, small changes.

We want people to ping when they honestly think it's an important change. Sometimes people ping too much ahead of their judgment, but that is not your case. My advice is always to use git 'blame' or Github to figure out who has been working on the code being changed and ping them. Sometimes Github's suggested reviewers are useful.

You can ping me on PRs you think are not getting reviewed and I'll review, and if I feel qualified and it's ready, can merge.

**Pull request closed without merging**
